Speed up markup removal

aikrahguzar opened the pull request:
Mostly to help with https://github.com/emacs-citar/citar/issues/659

The code still has rough edges but I think the overall approach is correct. It basically has one regexp for commands and accents and one for literals. One a match the command, accent or literal is looked up in an alist. The lookup probably can be sped up further by using a hash table but that doesn't seem it would affect anything (unless markup is very heavy).

With this I get,
```
ELISP> (benchmark-run-compiled 10 (let ((parsebib-TeX-markup-replace-alist nil)) (citar-cache--get-bibliography "bibfile" t)))
(6.233334512000001 55 4.1939954900000025)

ELISP> (benchmark-run-compiled 10 (citar-cache--get-bibliography "bibfile" t))
(7.2975037 62 4.9189581539999985)

```
with bibfile having about 1500 entries.

The code passes all the tests but probably has rough edges and for now lacks the excellent documentation everything around it has.

There is also the issue of propagating any changes in `parsebib-TeX-literal-replacement-alist` which needs to be addressed in some way.

bdarcus commented:
This looks really promising, @aikrahguzar!

BTW, a little git/github tip: generally easier/safer to do PRs off git feature branches, rather than master.

I saw recently a case where someone had their master/main branch accidentally wiped.

joostkremers commented:
Thanks! I'm on vacation, soni won't be able to look at the code for another two weeks, but I'll certainly check it out ass soon as I get back.

bdarcus commented:
@aikrahguzar - I ran the same benchmarks as you, but get different results.

```elisp
ELISP> (benchmark-run-compiled 10 (let ((parsebib-TeX-markup-replace-alist nil)) (citar-cache--get-bibliography "~/org/bib/academic.bib" t)))
(8.052586177 52 5.5369672070000036)

ELISP> (benchmark-run-compiled 10 (citar-cache--get-bibliography "~/org/bib/academic.bib" t))
(32.549427652 214 22.486489413)
```

aikrahguzar commented:
> Maybe something to do with how I'm loading it?

Can be if it is not a fresh session. What is the value for `parsebib-TeX-markup-replace-alist`? If you are evaluating the changes in a running session you will need to `eval-defun` the definition for `parsebib-TeX-markup-replace-alist`. Using `graf` on the definition for example will not change the value. 

Anyhow to get something comparable I tested with the bibliography that was posted in https://github.com/emacs-citar/citar/issues/659 and with it I get,
```
~ ❯ (benchmark-run-compiled 10 (citar-cache--get-bibliography "Downloads/bibfile.bib" t))
(30.804231123 194 18.045653951999995)

~ ❯ (benchmark-run-compiled 10 (let ((parsebib-TeX-markup-replace-alist nil)) (citar-cache--get-bibliography "Downloads/bibfile.bib" t)))
(26.210374116 161 14.870456689000008)

```
And I see in the message window.
```
Updating bibliography Downloads/bibfile.bib...done (3.048 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.032 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.125 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.027 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.121 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.052 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.135 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.038 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.129 seconds)
Updating bibliography Downloads/bibfile.bib...done (3.093 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.635 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.610 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.617 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.622 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.613 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.607 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.606 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.605 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.597 seconds)
Updating bibliography Downloads/bibfile.bib...done (2.694 seconds)
```

bdarcus commented:
> Can be if it is not a fresh session.

I probably loaded parsebib first, and then compiled and loaded your modified version.

aikrahguzar commented:
> I probably loaded parsebib first, and then compiled and loaded your modified version.

I think this is not going to have any effect. Since the function definition of `parsebib-clean-TeX-markup` is unchanged. It is the value of `parsebib-TeX-markup-replace-alist` that is changed i.e this pr aims to achieves the same result as before with much less fewer regexes. If `parsebib-TeX-markup-replace-alist` is already defined then the changed `defvar` statement is not going to chage anything unless you specifically `eval-defun` it.   

> But I can see your benchmarks, which is good enough for me! I was just trying to get a sense of what the difference was with the master version. E.g. I notice a delay with my smaller file; I assume it's noticeably improved here?

I hope so but it will be good if you try it out. You don't need to start a new session, just load the modified version and make sure the `parsebib-TeX-markup-replace-alist` is set to the new value i.e it should be an alist with exactly three entries.

bdarcus commented:
> You don't need to start a new session, just load the modified version and make sure the `parsebib-TeX-markup-replace-alist` is set to the new value i.e it should be an alist with exactly three entries.

Length is 72, so obviously not the right value. I'm not familiar with `eval-defun`.

aikrahguzar commented:
> > You don't need to start a new session, just load the modified version and make sure the `parsebib-TeX-markup-replace-alist` is set to the new value i.e it should be an alist with exactly three entries.
> 
> Length is 72, so obviously not the right value. I'm not familiar with `eval-defun`; so how do I set it correctly?

`eval-defun` is the non-evil version of `graf`. It should be on `SPC m e d`. You can also manually `setq`.

bdarcus commented:
Where does this stand ATM @aikrahguzar?

joostkremers commented:
> Where does this stand ATM @aikrahguzar?

I promised to take a look at the code, but haven't found time yet... Hopefully this weekend.

aikrahguzar commented:
> Where does this stand ATM @aikrahguzar?

The issue that needs to be figured out is to how to deal with literals. Right now they are defined in `parsebib-TeX-literal-replacement-alist` but the value of `parsebib-TeX-markup-replace-alist` depends on it. So if a user adds a literal/replacement pair to `parsebib-TeX-literal-replacement-alist`, `parsebib-TeX-markup-replace-alist` isn't changed to reflect that. So the question is how to expose this dependency to users and I don't have a very satisfactory solution. Some not so good ones I can think of are,
1) Make `parsebib-TeX-literal-replacement-alist` private and a `defconst` and declare that if users want to deal with more literals they should add another entry to `parsebib-TeX-markup-replace-alist`.
2) Make `parsebib-TeX-literal-replacement-alist` a defcustom with a setter that updates `parsebib-TeX-literal-replacement-alist` accordingly. But this only makes sense if we make `parsebib-TeX-literal-replacement-alist` private.
3) Make make `parsebib-TeX-literal-replacement-alist` private and provide a function to update it.

joostkremers commented:
> The issue that needs to be figured out is to how to deal with literals.

Actually, I'm not sure this is a serious issue, given that this is a developer-facing package. (It would be a serious issue for a user-facing package, I agree.)

> [...]
> So the question is how to expose this dependency to users 

Since users are developers, I think decent documentation should be sufficient here.

>     1. Make `parsebib-TeX-literal-replacement-alist` private and a `defconst` and declare that if users want to deal with more literals they should add another entry to `parsebib-TeX-markup-replace-alist`.

But we're trying to streamline the regexes in `parsebib-TeX-markup-replace-alist`, so telling users to add entries to it doesn't sound like a good idea.

>     2. Make `parsebib-TeX-literal-replacement-alist` a defcustom with a setter that updates `parsebib-TeX-literal-replacement-alist` accordingly. But this only makes sense if we make `parsebib-TeX-literal-replacement-alist` private.

Since I prefer parsebib to be customisable by `let`-binding the relevant variables around calls to parsebib, a defcustom with a setter is not an option either, I think. Besides, like I said, this isn't a user-facing package, so I don't think defcustoms are the way to go in general.

Note that even in the current version, `parsebib-TeX-markup-replace-alist` is already not very user-friendly, because its value is computed. If a package such as citar or Ebib would want to make it customisable, it would have to wrap it into a `defcustom` and provide a setter. This PR doesn't make this worse.

>     3. Make make `parsebib-TeX-literal-replacement-alist` private and provide a function to update it.

If there is a desire from packages using parsebib to customise the replacements, this would be a viable option, I think.

I don't know if there is any interest in making TeX markup replacements user-customisable, though. I'm not really planning to do so in Ebib, but of course I don't know about other packages.

If there isn't, I think it's enough to document the relation between `parsebib-TeX-literal-replacement-alist` and `parsebib-TeX-markup-replace-alist` and that in general the latter cannot be customised by adding elements to it. Rather, it should be recomputed and that developers should look at the source code to find out how to do that.

joostkremers commented:
BTW, Perhaps we should rename `parsebib-TeX-markup-replace-alist` to `parsebib-TeX-markup-replacement-alist` for consistency with the other variables.

bdarcus commented:
> BTW, Perhaps we should rename `parsebib-TeX-markup-replace-alist` to `parsebib-TeX-markup-replacement-alist` for consistency with the other variables.

So basically, with this change, and some docstring tweaks, this is RTM @joostkremers?

I agree we shouldn't worry too much about customization, and to just require developers to using let-binding.

BTW, once this is merged, and probably also your plan to change the fields to an include list, I'll bump the doom module (which I maintain!).

aikrahguzar commented:
> Actually, I'm not sure this is a serious issue, given that this is a developer-facing package. (It would be a serious issue for a user-facing package, I agree.)

This distinction makes complete sense and one I wasn't thinking of.

But one thing that comes to mind about the developer experience. What about exchanging the car and cdr of alist entries? We loop over entries `parsebib-TeX-markup-replace-alist` so the order doesn't really matter but with the replacement coming first, the way to modify it if an entry is added to `parsebib-TeX-literal-replacement-alist` is `(setf (alist-get 'parsebib--replace-literal) ...)` while with the current order the way to do it is somewhat messy. 

On the other hand, this causes inconsistency with literal and command alists and I don't see it being useful for much else than the literal entry.

aikrahguzar commented:
I changed the order of the car/cdr in `parsebib-TeX-markup-replacement-alist`  (which is now renamed) and added some documentation to the new `defvar`s including how to adjust the mark replacement alist when a new literal is added.

bdarcus commented:
So this is no longer a draft PR then?

aikrahguzar commented:
> So this is no longer a draft PR then?

Yes, I just changed it as ready for review. But the documentation is of the new stuff is still light by rest of parsebib standards and suggestions for changes are welcome too.

bdarcus commented:
I get a compilation warning:

> parsebib.el:342:2: Warning: defvar `parsebib-TeX-literal-replacement-alist'
    docstring has wrong usage of unescaped single quotes (use \= or different
    quoting)

aikrahguzar commented:
> I get a compilation warning:

I see that the docstring rendering is indeed thrown off and there is a stray quote at the end of that bit of code.

> Changing line 363 to this resolves it.
> 
> ```emacs-lisp
>     (alist-get \\'parsebib--replace-literal
> ```

But this results in `parsebib--replace-literal` not being quoted. I wonder what is the correct way to include a quote snippet in a docstring. For now the best I seem to be able to do is to replace `'parsebib--replace-literal` by the explicit `(quote parsebib--replace-literal)`

bdarcus commented:
> For now the best I seem to be able to do is to replace `'parsebib--replace-literal` by the explicit `(quote parsebib--replace-literal)`

There's still correct; no? And it's just a docstring.

Better that than a compiler warning.

So I tested this, and even with my smaller ~1000 entry bib file, initial parsing is noticeably quicker.

LGTM!

bdarcus commented:
@joostkremers?

joostkremers commented on parsebib.el:
```diff
@@ -355,13 +386,38 @@ simply replace the match) or a function (the match will be
 string).  Earlier elements are evaluated before later ones, so if
 one string is a subpattern of another, the second must appear
-later (e.g. \"''\" is before \"'\").")
+later (e.g. \"\" is before \"'\").")
```
This change is not correct. It really should say `''` and not `"`.

joostkremers commented on parsebib.el:
```diff
@@ -240,113 +240,144 @@ trailing space will be included in the overall match."
+(defvar parsebib-TeX-markup-replacement-alist
+  `((parsebib--replace-command-or-accent
+     . ,(rx "\\" (group-n 1 (or (1+ letter) nonl))
+
```
Why the empty line?

joostkremers commented on parsebib.el:
```diff
@@ -240,113 +240,144 @@ trailing space will be included in the overall match."
+
+will ensure that LITERAL gets replaced with REPLACEMENT.")
+
+(defvar parsebib-TeX-markup-replacement-alist
```
The `defvar` should have a doc string.

joostkremers commented:
> @joostkremers?

Yeah, I know, I'm sorry. I haven't found any time yet to look at this in more detail. I took a quick look and made a few comments, but it's not enough to really understand what the new code is doing and how it's doing it. I'll try to find time this week.

BTW, @aikrahguzar, with a view to contributing parsebib to GNU Elpa, do you have a copyright assignment on file? (It's possible non-GNU Elpa is the best we can hope for anyway, so not a problem if not.)

joostkremers commented on parsebib.el:
```diff
@@ -240,113 +240,144 @@ trailing space will be included in the overall match."
+
+COMMAND is the name of a TeX or LaTeX command (without
+backslash).  REPLACEMENT can be either a string or a
+function. If it is a function, it is passed the first
```
There should be two spaces after a full stop. (Use checkdoc to look for issues with doc strings.)

aikrahguzar replied:
Checkdoc does turn up anything for me but I fixed the cases I found. Hopefully all.

joostkremers commented on parsebib.el:
```diff
@@ -240,113 +240,144 @@ trailing space will be included in the overall match."
+`(cl-callf (lambda (regex) (rx (or LITERAL (regexp regex))))
+    (alist-get (quote parsebib--replace-literal)
+               parsebib-TeX-markup-replacement-alist))'
+
```
I'd prefer two spaces before the code snippet and no `...':

```
  (cl-callf (lambda (regex) (rx (or LITERAL (regexp regex))))
      (alist-get (quote parsebib--replace-literal)
                 parsebib-TeX-markup-replacement-alist))
```

aikrahguzar replied:
Done

joostkremers commented on parsebib.el:
```diff
@@ -240,113 +240,144 @@ trailing space will be included in the overall match."
+(defvar parsebib-TeX-markup-replacement-alist
+  `((parsebib--replace-command-or-accent
+     . ,(rx "\\" (group-n 1 (or (1+ letter) nonl))
+
```
I don't think an empty line here is necessary, unless you have a particular reason for it?

aikrahguzar replied:
Removed. It was accidental.

joostkremers commented on parsebib.el:
```diff
@@ -355,13 +386,38 @@ simply replace the match) or a function (the match will be
 string).  Earlier elements are evaluated before later ones, so if
 one string is a subpattern of another, the second must appear
-later (e.g. \"''\" is before \"'\").")
+later (e.g. \"\" is before \"'\").")
```
Probably an accident: the two single quotes should not be deleted.

aikrahguzar replied:
It was, now fixed.

bdarcus commented:
> BTW, @aikrahguzar, with a view to contributing parsebib to GNU Elpa, do you have a copyright assignment on file?

I forgot about this.

@aikrahguzar I"m guessing you don't, but it's pretty easy to do, particularly if you're not employed as a programmer. You basically just send an email with some info.

aikrahguzar commented on parsebib.el:
```diff
@@ -240,113 +240,144 @@ trailing space will be included in the overall match."
+
+will ensure that LITERAL gets replaced with REPLACEMENT.")
+
+(defvar parsebib-TeX-markup-replacement-alist
```
I made some adjustments.

aikrahguzar commented:
> > BTW, @aikrahguzar, with a view to contributing parsebib to GNU Elpa, do you have a copyright assignment on file?
> 
> I forgot about this.
> 
> @aikrahguzar I"m guessing you don't, but it's easy to do, particularly if you're not employed as a programmer. You basically just send an email with some info, they send you back a PDF form to sign in a few days (should be no more than 5 business days I think), and then you send back the signed and scanned form.
> 
> https://orgmode.org/request-assign-future.txt
> 
> That will then also allow you to contribute to Emacs in the future, if you wish.

I don't have it and I am employed by a university so bureaucracy will be necessary. But it is the only thing holding `parsebib` from getting into elpa, let me know and I will try to get the copyright assignment done.

aikrahguzar commented:
> I took a quick look and made a few comments, but it's not enough to really understand what the new code is doing and how it's doing it. I'll try to find time this week.

I added more commentary around the new command regexp and how the command replacement works which are the gnarly parts of the new code so hopefully that helps.

bdarcus commented:
> I am employed by a university so bureaucracy will be necessary.

I thought this might be the case when I did this, but FSF told me "If you are not employed to program, then we do not need an employer disclaimer."

joostkremers commented:
> Since I just have a typical academic position, that was the case with me (and with Joost, I believe).

No, not anymore. But yes, the procedure is rather painless. The employer disclaimer is only required if you are a programmer. If your contract doesn't mention programming or software development, you're fine.

joostkremers commented:
I finally got round to taking a closer look and merging this PR. @aikrahguzar thank you very much for writing the code!

kaushalmodi commented:
Hello all, can the `rx` code in this PR be rewritten in Emacs 26.3 compatible fashion? See https://github.com/joostkremers/parsebib/issues/26.

joostkremers commented:
> Hello all, can the `rx` code in this PR be rewritten in Emacs 26.3 compatible fashion? See #26.

Ugh, yeah, that needs to be fixed... Suggestions on how to fix this are very welcome. Otherwise suggestions on how to best run an Emacs 26 instance without interfering with the existing Emacs 28 installation...

bdarcus commented:
For the future, you could add a github action CI to check all PRs?

We somewhat recently added one for citar.

kaushalmodi commented:
> Otherwise suggestions on how to best run an Emacs 26 instance without interfering with the existing Emacs 28 installation...

If you build Emacs locally, clone the Emacs repo, check out the emacs-26 branch and run configure with `--prefix` pointing to a separate directory where you want to install Emacs 26. That's how I do it. 

But for debugging most of such failures, I just look at the failure logs on GitHub Actions. Here's my GHA configuration for `ox-hugo`: https://github.com/kaushalmodi/ox-hugo/blob/main/.github/workflows/test.yml

---

> Suggestions on how to fix this are very welcome.

Unfortunately, I haven't used `rx` enough. I tried to use it few times, but every time I want to use it again, I forget its special syntax, and I fall back to using plain old regexp.
